File: go/openai/responses/decode.go
package responses

import (
	"fmt"

	"github.com/oxa-protocol/oxa/go/ir"
)

// DecodeRequest converts a Responses wire request to the IR (face -> IR).
// Semantic unmappables are losses, never errors; errors are reserved for
// structural type violations of known fields (spec/02 s4).
func DecodeRequest(wire *Request, opts ...Option) (*ir.Request, []ir.Loss, error) {
	if wire == nil {
		return nil, nil, fmt.Errorf("responses: nil request")
	}
	var losses []ir.Loss
	for _, field := range []struct {
		path    string
		field   string
		present bool
		detail  string
	}{
		{"metadata", "metadata", len(wire.Metadata) > 0,
			"Responses request metadata has no IR equivalent in v1."},
		{"text.verbosity", "verbosity", wire.Text != nil && wire.Text.Verbosity != nil,
			"Responses output verbosity has no IR equivalent in v1."},
		{"text.format", "format", wire.Text != nil && wire.Text.Format != nil,
			"Responses text output format has no IR equivalent in v1."},
		{"reasoning", "reasoning", wire.Reasoning != nil,
			"Responses reasoning effort configuration has no IR equivalent in v1."},
		{"parallel_tool_calls", "parallel_tool_calls", wire.ParallelToolCalls != nil,
			"Responses parallel tool calls have no IR equivalent in v1."},
	} {
		if field.present {
			losses = append(losses, loss(field.path, field.field, ir.LossUnmappedField, field.detail))
		}
	}

	o := newOptions(opts...)
	req := &ir.Request{Model: o.models.Map(wire.Model)}

	// N-R-1: instructions render as the first system block; system items in
	// the input array follow in document order.
	if wire.Instructions != nil {
		req.System = append(req.System, ir.SystemBlock{Text: *wire.Instructions})
	}

	for i, tool := range wire.Tools {
		if tool.Type != "function" {
			losses = append(losses, loss(
				fmt.Sprintf("tools[%d]", i), "type", ir.LossUnsupportedSemantic,
				fmt.Sprintf("Responses tool type %q has no IR equivalent", tool.Type),
			))
			continue
		}
		req.Tools = append(req.Tools, ir.Tool{
			Name:        tool.Name,
			Description: tool.Description,
			InputSchema: tool.Parameters,
		})
		if tool.Strict != nil {
			losses = append(losses, loss(
				fmt.Sprintf("tools[%d].strict", i), "strict", ir.LossUnmappedField,
				"Responses function tool strict mode has no IR equivalent in v1.",
			))
		}
	}
	choice, choiceLosses := decodeToolChoice(wire.ToolChoice)
	req.ToolChoice = choice
	losses = append(losses, choiceLosses...)

	// N-R-2: the string shorthand is one user message with one text block.
	if wire.Input.Text != nil {
		req.Messages = append(req.Messages, ir.Message{
			Role:    ir.RoleUser,
			Content: []ir.Block{ir.TextBlock{Text: *wire.Input.Text}},
		})
	}

	i := 0
	for i < len(wire.Input.Items) {
		item := wire.Input.Items[i]
		switch {
		case item.Type == "" || item.Type == "message":
			switch item.Role {
			case "system":
				content, contentLosses, err := decodeItemContent(item.Content, fmt.Sprintf("input[%d].content", i))
				if err != nil {
					return nil, nil, err
				}
				for j, block := range content {
					if text, ok := block.(ir.TextBlock); ok {
						req.System = append(req.System, ir.SystemBlock{Text: text.Text})
						continue
					}
					losses = append(losses, loss(
						fmt.Sprintf("input[%d].content[%d]", i, j), "content", ir.LossUnsupportedSemantic,
						fmt.Sprintf("IR %T cannot be rendered in the IR system field", block),
					))
				}
				losses = append(losses, contentLosses...)
				i++
			case "user":
				content, contentLosses, err := decodeItemContent(item.Content, fmt.Sprintf("input[%d].content", i))
				if err != nil {
					return nil, nil, err
				}
				losses = append(losses, contentLosses...)
				req.Messages = append(req.Messages, ir.Message{Role: ir.RoleUser, Content: content})
				i++
			case "assistant":
				merged, next, runLosses, err := decodeAssistantRun(wire.Input.Items, i)
				if err != nil {
					return nil, nil, err
				}
				if merged != nil {
					req.Messages = append(req.Messages, *merged)
				}
				losses = append(losses, runLosses...)
				i = next
			default:
				return nil, nil, fmt.Errorf("responses: input[%d]: unknown role %q", i, item.Role)
			}
		case item.Type == "function_call":
			merged, next, runLosses, err := decodeAssistantRun(wire.Input.Items, i)
			if err != nil {
				return nil, nil, err
			}
			if merged != nil {
				req.Messages = append(req.Messages, *merged)
			}
			losses = append(losses, runLosses...)
			i = next
		case item.Type == "function_call_output":
			merged, next, runLosses, err := decodeOutputRun(wire.Input.Items, i)
			if err != nil {
				return nil, nil, err
			}
			req.Messages = append(req.Messages, *merged)
			losses = append(losses, runLosses...)
			i = next
		default:
			losses = append(losses, loss(
				fmt.Sprintf("input[%d]", i), "type", ir.LossUnsupportedSemantic,
				fmt.Sprintf("Responses input item type %q has no IR equivalent", item.Type),
			))
			i++
		}
	}
	if len(req.Messages) == 0 {
		return nil, nil, fmt.Errorf("responses: request carries no conversation input")
	}
	req.Params = ir.Params{
		Temperature: wire.Temperature,
		TopP:        wire.TopP,
		MaxTokens:   wire.MaxOutputTokens,
	}
	return req, losses, nil
}

// decodeAssistantRun converts a maximal run of assistant message items and
// function_call items into one IR assistant message (N-R-5): text blocks
// first (in document order), then ToolUseBlocks.
func decodeAssistantRun(items []InputItem, start int) (*ir.Message, int, []ir.Loss, error) {
	var blocks []ir.Block
	var losses []ir.Loss
	i := start
	for ; i < len(items); i++ {
		item := items[i]
		if item.Type == "function_call" {
			input, err := wrapToolArguments(item.Arguments)
			if err != nil {
				return nil, 0, nil, fmt.Errorf("responses: input[%d].arguments: %w", i, err)
			}
			blocks = append(blocks, ir.ToolUseBlock{ID: item.CallID, Name: item.Name, Input: input})
			continue
		}
		if item.Type != "" && item.Type != "message" {
			break
		}
		if item.Role != "assistant" {
			break
		}
		content, contentLosses, err := decodeItemContent(item.Content, fmt.Sprintf("input[%d].content", i))
		if err != nil {
			return nil, 0, nil, err
		}
		losses = append(losses, contentLosses...)
		blocks = append(blocks, content...)
	}
	if len(blocks) == 0 {
		return nil, i, losses, nil
	}
	return &ir.Message{Role: ir.RoleAssistant, Content: blocks}, i, losses, nil
}

// decodeOutputRun converts a maximal run of consecutive function_call_output
// items into one IR user message of ToolResultBlocks (N-R-6, INV-4).
func decodeOutputRun(items []InputItem, start int) (*ir.Message, int, []ir.Loss, error) {
	content := make([]ir.Block, 0)
	i := start
	for ; i < len(items) && items[i].Type == "function_call_output"; i++ {
		content = append(content, ir.ToolResultBlock{
			ToolUseID: items[i].CallID,
			Content:   []ir.Block{ir.TextBlock{Text: items[i].Output}},
		})
	}
	return &ir.Message{Role: ir.RoleUser, Content: content}, i, nil, nil
}

// decodeToolChoice maps the Responses tool_choice forms to the IR modes
// (N-R-9): required and IR mode any are equivalent, loss-free both ways.
func decodeToolChoice(value any) (*ir.ToolChoice, []ir.Loss) {
	if value == nil {
		return nil, nil
	}
	if choice, ok := value.(string); ok {
		switch choice {
		case "auto", "none":
			return &ir.ToolChoice{Mode: choice}, nil
		case "required":
			return &ir.ToolChoice{Mode: "any"}, nil
		default:
			return nil, []ir.Loss{loss("tool_choice", "tool_choice", ir.LossUnsupportedSemantic,
				fmt.Sprintf("Responses tool_choice %q has no IR equivalent", choice))}
		}
	}
	var kind, name string
	switch choice := value.(type) {
	case ToolChoiceWire:
		kind, name = choice.Type, choice.Name
	case *ToolChoiceWire:
		if choice == nil {
			return nil, nil
		}
		kind, name = choice.Type, choice.Name
	case map[string]any:
		kind, _ = choice["type"].(string)
		name, _ = choice["name"].(string)
	default:
		return nil, []ir.Loss{loss("tool_choice", "tool_choice", ir.LossUnsupportedSemantic,
			"Responses tool_choice has no IR equivalent")}
	}
	if kind == "function" && name != "" {
		return &ir.ToolChoice{Mode: "tool", Name: name}, nil
	}
	return nil, []ir.Loss{loss("tool_choice", "tool_choice", ir.LossUnsupportedSemantic,
		"only named function Responses tool_choice values are supported")}
}

// encodeToolChoice renders the IR tool-choice modes in Responses form
// (N-R-9).
func encodeToolChoice(choice *ir.ToolChoice) (any, *ir.Loss) {
	if choice == nil {
		return nil, nil
	}
	switch choice.Mode {
	case "auto", "none":
		return choice.Mode, nil
	case "any":
		return "required", nil
	case "tool":
		if choice.Name == "" {
			return nil, ptrLoss(loss("tool_choice", "tool_choice", ir.LossUnsupportedSemantic,
				"IR named tool choice has no function name"))
		}
		return ToolChoiceWire{Type: "function", Name: choice.Name}, nil
	default:
		return nil, ptrLoss(loss("tool_choice", "tool_choice", ir.LossUnsupportedSemantic,
			fmt.Sprintf("IR tool_choice mode %q has no Responses equivalent", choice.Mode)))
	}
}

// DecodeResponse converts a Responses wire response to the IR (face -> IR).
// Envelope fields (object, item ids/status) are exempt from losses
// (vectors/README.md loss conventions); usage.total_tokens is derived.
func DecodeResponse(wire *Response, opts ...Option) (*ir.Response, []ir.Loss, error) {
	if wire == nil {
		return nil, nil, fmt.Errorf("responses: nil response")
	}
	var losses []ir.Loss
	var blocks []ir.Block
	hasToolUse := false
	for i, item := range wire.Output {
		switch item.Type {
		case "message":
			for j, part := range item.Content {
				if part.Type != "output_text" {
					losses = append(losses, loss(
						fmt.Sprintf("output[%d].content[%d]", i, j), "type", ir.LossUnsupportedSemantic,
						fmt.Sprintf("Responses output content type %q has no IR equivalent", part.Type),
					))
					continue
				}
				if len(part.Annotations) > 0 {
					losses = append(losses, loss(
						fmt.Sprintf("output[%d].content[%d].annotations", i, j), "annotations", ir.LossUnmappedField,
						"Responses output annotations have no IR equivalent in v1.",
					))
				}
				blocks = append(blocks, ir.TextBlock{Text: part.Text})
			}
		case "function_call":
			input, err := wrapToolArguments(item.Arguments)
			if err != nil {
				return nil, nil, fmt.Errorf("responses: output[%d].arguments: %w", i, err)
			}
			blocks = append(blocks, ir.ToolUseBlock{ID: item.CallID, Name: item.Name, Input: input})
			hasToolUse = true
		default:
			losses = append(losses, loss(
				fmt.Sprintf("output[%d]", i), "type", ir.LossUnsupportedSemantic,
				fmt.Sprintf("Responses output item type %q has no IR equivalent", item.Type),
			))
		}
	}
	stop, stopLosses, err := decodeStatus(wire, hasToolUse)
	if err != nil {
		return nil, nil, err
	}
	losses = append(losses, stopLosses...)
	o := newOptions(opts...)
	resp := &ir.Response{
		ID:         wire.ID,
		Model:      o.models.Map(wire.Model),
		Content:    blocks,
		StopReason: stop,
	}
	if wire.Usage != nil {
		resp.Usage = ir.Usage{
			InputTokens:  wire.Usage.InputTokens,
			OutputTokens: wire.Usage.OutputTokens,
		}
	}
	return resp, losses, nil
}

// decodeStatus maps the Responses status/incomplete_details/error triple to
// the IR stop reason (N-R-11).
func decodeStatus(wire *Response, hasToolUse bool) (ir.StopReason, []ir.Loss, error) {
	if wire.Error != nil {
		return ir.StopOther, []ir.Loss{loss(
			"error", "error", ir.LossUnsupportedSemantic,
			fmt.Sprintf("failed Responses response carries error %q: %s", wire.Error.Code, wire.Error.Message),
		)}, nil
	}
	switch wire.Status {
	case "completed":
		if hasToolUse {
			return ir.StopToolUse, nil, nil
		}
		return ir.StopEndTurn, nil, nil
	case "incomplete":
		reason := ""
		if wire.IncompleteDetails != nil {
			reason = wire.IncompleteDetails.Reason
		}
		if reason == "max_output_tokens" {
			return ir.StopMaxTokens, nil, nil
		}
		return ir.StopOther, []ir.Loss{loss(
			"incomplete_details.reason", "reason", ir.LossUnmappedValue,
			fmt.Sprintf("Responses incomplete_details reason %q has no IR equivalent", reason),
		)}, nil
	case "failed":
		return ir.StopOther, []ir.Loss{loss(
			"error", "error", ir.LossUnsupportedSemantic,
			"failed Responses response carries no error object",
		)}, nil
	default:
		return "", nil, fmt.Errorf("responses: status %q has no IR equivalent", wire.Status)
	}
}

File: python/src/oxa/openai/chatcompletions/decode.py
"""Chat Completions face → IR conversions (spec/10 §4)."""

from __future__ import annotations

from typing import Any

from oxa.ir import (
    LOSS_UNMAPPED_FIELD,
    LOSS_UNMAPPED_VALUE,
    LOSS_UNSUPPORTED_SEMANTIC,
    ROLE_ASSISTANT as IR_ROLE_ASSISTANT,
    ROLE_USER as IR_ROLE_USER,
    STOP_END_TURN,
    STOP_MAX_TOKENS,
    STOP_OTHER,
    STOP_REFUSAL,
    STOP_TOOL_USE,
    Block,
    Loss,
    Message as IrMessage,
    Params,
    Request,
    Response,
    SystemBlock,
    TextBlock,
    Tool,
    Usage,
)
from oxa.modelmap import Table
from oxa.openai.chatcompletions.constants import (
    FINISH_REASON_CONTENT_FILTER,
    FINISH_REASON_LENGTH,
    FINISH_REASON_STOP,
    FINISH_REASON_TOOL_CALLS,
    ROLE_ASSISTANT,
    ROLE_SYSTEM,
    ROLE_TOOL,
    ROLE_USER,
    TOOL_TYPE_FUNCTION,
)
from oxa.openai.chatcompletions.normalize import (
    content_system,
    decode_content,
    decode_tool_calls,
    decode_tool_choice,
    decode_tool_result_run,
    loss,
)


def decode_request(
    wire: dict[str, Any],
    raw_text: str = "",
    table: Table | None = None,
) -> tuple[Request, list[Loss]]:
    """Converts a Chat Completions wire request to the IR (face → IR)."""
    losses: list[Loss] = []

    unmapped_fields = [
        ("parallel_tool_calls", "Chat Completions parallel tool calls have no IR equivalent in v1."),
        ("functions", "legacy Chat Completions functions have no IR equivalent in v1."),
        ("function_call", "legacy Chat Completions function_call has no IR equivalent in v1."),
        ("response_format", "Chat Completions response_format has no IR equivalent in v1."),
        ("logprobs", "Chat Completions log-probability sampling has no IR equivalent in v1."),
        ("top_logprobs", "Chat Completions log-probability sampling has no IR equivalent in v1."),
        ("metadata", "Chat Completions request metadata has no IR equivalent in v1."),
    ]
    for name, detail in unmapped_fields:
        if wire.get(name) is not None:
            losses.append(loss(name, name, LOSS_UNMAPPED_FIELD, detail))

    model = str(wire.get("model", ""))
    if table is not None:
        model = table.map(model)

    tools: list[Tool] | None = None
    if "tools" in wire and wire["tools"] is not None:
        tools = []
        for index, tool in enumerate(wire["tools"]):
            kind = tool.get("type", "")
            if kind != TOOL_TYPE_FUNCTION:
                losses.append(
                    loss(
                        f"tools[{index}]",
                        "type",
                        LOSS_UNSUPPORTED_SEMANTIC,
                        f"Chat Completions tool type {kind!r} has no IR equivalent",
                    )
                )
                continue
            func = tool.get("function", {})
            desc = func.get("description")
            tools.append(
                Tool(
                    name=func.get("name", ""),
                    description=desc if desc else None,
                    input_schema=func.get("parameters", {}),
                )
            )

    tool_choice, tc_losses = decode_tool_choice(wire.get("tool_choice"))
    losses.extend(tc_losses)

    system: list[SystemBlock] = []
    messages: list[IrMessage] = []
    raw_messages = wire.get("messages", [])

    index = 0
    while index < len(raw_messages):
        msg = raw_messages[index]
        role = msg.get("role", "")
        if role == ROLE_TOOL:
            merged, next_idx, result_losses = decode_tool_result_run(raw_messages, index)
            messages.append(merged)
            losses.extend(result_losses)
            index = next_idx
            continue

        content_path = f"messages[{index}].content"
        content, content_losses = decode_content(msg.get("content"), content_path)
        losses.extend(content_losses)

        if role != ROLE_SYSTEM and not content:
            content = [TextBlock(text="")]

        if role == ROLE_SYSTEM:
            sys_blocks, sys_losses = content_system(content, content_path)
            system.extend(sys_blocks)
            losses.extend(sys_losses)
        elif role == ROLE_USER:
            messages.append(IrMessage(role=IR_ROLE_USER, content=content))
        elif role == ROLE_ASSISTANT:
            tool_calls = msg.get("tool_calls")
            if tool_calls is not None:
                calls_path = f"messages[{index}].tool_calls"
                tool_blocks, tool_losses = decode_tool_calls(tool_calls, calls_path)
                if msg.get("content") is None and tool_blocks:
                    content.clear()
                content.extend(tool_blocks)
                losses.extend(tool_losses)
            messages.append(IrMessage(role=IR_ROLE_ASSISTANT, content=content))
        else:
            raise ValueError(f"chatcompletions: messages[{index}]: unknown role {role!r}")

        if "function_call" in msg and msg["function_call"] is not None:
            losses.append(
                loss(
                    f"messages[{index}].function_call",
                    "function_call",
                    LOSS_UNMAPPED_FIELD,
                    "legacy Chat Completions function_call has no IR equivalent",
                )
            )
        index += 1

    if not messages:
        raise ValueError("chatcompletions: request carries no conversation messages")

    stop = wire.get("stop")
    stop_sequences: list[str] | None = None
    if isinstance(stop, str):
        stop_sequences = [stop] if stop else None
    elif isinstance(stop, list):
        stop_sequences = [s for s in stop if s] if stop else None

    max_tokens = wire.get("max_tokens")
    if max_tokens is None:
        max_tokens = wire.get("max_completion_tokens")

    params: Params | None = None
    if (
        wire.get("temperature") is not None
        or wire.get("top_p") is not None
        or max_tokens is not None
        or stop_sequences is not None
    ):
        params = Params(
            temperature=wire.get("temperature"),
            top_p=wire.get("top_p"),
            max_tokens=max_tokens,
            stop_sequences=stop_sequences,
        )

    return (
        Request(
            model=model,
            messages=messages,
            system=system,
            tools=tools if tools else None,
            tool_choice=tool_choice,
            params=params,
        ),
        losses,
    )


def decode_response(
    wire: dict[str, Any],
    raw_text: str = "",
    table: Table | None = None,
) -> tuple[Response, list[Loss]]:
    """Converts a Chat Completions wire response to the IR (face → IR)."""
    choices = wire.get("choices")
    if not choices or not isinstance(choices, list):
        raise ValueError("chatcompletions: choices array is required and must not be empty")

    choice = choices[0]
    msg = choice.get("message", {})
    content, losses = decode_content(msg.get("content"), "choices[0].message.content")

    tool_calls = msg.get("tool_calls")
    if tool_calls is not None:
        tool_blocks, tool_losses = decode_tool_calls(tool_calls, "choices[0].message.tool_calls")
        if msg.get("content") is None and tool_blocks:
            content.clear()
        content.extend(tool_blocks)
        losses.extend(tool_losses)

    stop_reason, finish_loss = decode_finish_reason(choice.get("finish_reason"))
    if finish_loss is not None:
        losses.append(finish_loss)

    model = str(wire.get("model", ""))
    if table is not None:
        model = table.map(model)

    usage_raw = wire.get("usage", {})
    usage = Usage(
        input_tokens=int(usage_raw.get("prompt_tokens", 0)),
        output_tokens=int(usage_raw.get("completion_tokens", 0)),
    )

    return (
        Response(
            id=str(wire.get("id", "")),
            model=model,
            content=content,
            stop_reason=stop_reason,
            usage=usage,
        ),
        losses,
    )


def decode_finish_reason(finish: str | None) -> tuple[str, Loss | None]:
    if not finish:
        raise ValueError("chatcompletions: choices[0].finish_reason is missing")
    if finish == FINISH_REASON_STOP:
        return STOP_END_TURN, None
    if finish == FINISH_REASON_LENGTH:
        return STOP_MAX_TOKENS, None
    if finish == FINISH_REASON_CONTENT_FILTER:
        return STOP_REFUSAL, None
    if finish == FINISH_REASON_TOOL_CALLS:
        return STOP_TOOL_USE, None
    return (
        STOP_OTHER,
        loss(
            "choices[0].finish_reason",
            "finish_reason",
            LOSS_UNMAPPED_VALUE,
            f"Chat Completions finish_reason {finish!r} has no IR equivalent",
        ),
    )
